frontend/ui/widget/qrwidget.lua

```diff
@@ -4,34 +4,43 @@ QRWidget shows a QR code for a given text.
 
 local ImageWidget = require("ui/widget/imagewidget")
 local RenderImage = require("ui/renderimage")
-local Widget = require("ui/widget/widget")
 local logger = require("logger")
 local qrencode = require("qrencode")
 
-local QRWidget = Widget:new{
+local QRWidget = ImageWidget:extend{
     scale_factor = 0,
     text = ""
     -- see ImageWidget for other options.
 }
 
 function QRWidget:init()
-    local ok, ret = qrencode.qrcode(self.text)
+    local ok, ret = qrencode.qrcode(self.text:sub(1, 2953))
     if not ok then
         logger.info("QRWidget: failed to generate QR code.")
         return
     else
         -- We generate pbm data
-        local qr = {"P1\n", #ret[1], " ", #ret, "\n"}
+        local qr = {"P1\n", 2*#ret[1], " ", 2*#ret, "\n"}
         for _, col in ipairs(ret) do
-            for _, lgn in ipairs(col) do
-                table.insert(qr, lgn > 0 and "1 " or "0 ")
+            for i = 1, 2 do
+                for _, lgn in ipairs(col) do
+                    table.insert(qr, lgn > 0 and "1 1 " or "0 0 ")
+                end
+                table.insert(qr, "\n")
```

Comment:
I'm a bit fuzzy on the PBM format, but doesn't that introduce newlines after each "output" pixel, instead of after each scanline?

Reply:
The examples on [Wikipedia](https://en.wikipedia.org/wiki/Netpbm#PBM_example) put newlines only after each scanline, but I'm not sure at all !

Reply:
That was my understanding, too, but I'll admit I'd need to actually run the code myself to check if that's what happens here, and I'm not on my dev box, so I can't ;).